### PR TITLE
closes #398 cli for `assert_function()`

### LINKED ISSUE #398: `assert_function()`

### PR DESCRIPTION
**Pull request by ddsjoberg**
Thank you for your Pull Request! We have developed this task checklist from the [Development Process Guide](https://pharmaverse.github.io/admiral/dev/CONTRIBUTING.html) to help with the final steps of the process. Completing the below tasks helps to ensure our reviewers can maximize their time on your code as well as making sure the admiral codebase remains robust and consistent.   

Please check off each taskbox as an acknowledgment that you completed the task or check off that it is not relevant to your Pull Request. This checklist is part of the Github Action workflows and the Pull Request will not be merged into the `main` branch until you have checked off each task.

- [x] Place Closes #398 into the beginning of your Pull Request Title (Use Edit button in top-right if you need to update)
- [x] Code is formatted according to the [tidyverse style guide](https://style.tidyverse.org/). Run `styler::style_file()` to style R and Rmd files
- [x] Updated relevant unit tests or have written new unit tests, which should consider realistic data scenarios and edge cases, e.g. empty datasets, errors, boundary cases etc. - See [Unit Test Guide](https://pharmaverse.github.io/admiraldev/articles/unit_test_guidance.html#tests-should-be-robust-to-cover-realistic-data-scenarios)
- [x] If you removed/replaced any function and/or function parameters, did you fully follow the [deprecation guidance](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#deprecation)?
- [x] Update to all relevant roxygen headers and examples, including keywords and families. Refer to the [categorization of functions](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#categorization-of-functions) to tag appropriate keyword/family.
- [x] Run `devtools::document()` so all `.Rd` files in the `man` folder and the `NAMESPACE` file in the project root are updated appropriately
- [x] Address any updates needed for vignettes and/or templates
- [x] Update `NEWS.md` under the header `# admiral (development version)` if the changes pertain to a user-facing function (i.e. it has an `@export` tag) or documentation aimed at users (rather than developers)
- [x] Build admiral site `pkgdown::build_site()` and check that all affected examples are displayed correctly and that all new functions occur on the "[Reference](https://pharmaverse.github.io/admiraldev/reference/index.html)" page. 
- [x] Address or fix all lintr warnings and errors - `lintr::lint_package()`
- [x] Run `R CMD check` locally and address all errors and warnings - `devtools::check()`
- [x] Link the issue in the Development Section on the right hand side.
- [x] Address all merge conflicts and resolve appropriately
- [x] Pat yourself on the back for a job well done! Much love to your accomplishment!

### CODE REVIEW

File: R/assertions.R [hunk 1]
```diff
@@ -1047,45 +1047,58 @@ assert_has_variables <- function(dataset, required_vars) {
...
     )
-    abort(err_msg)
   }
   if (!is.null(params)) {
     if ("..." %in% names(formals(arg))) {
       return(invisible(arg))
     }
     is_param <- params %in% names(formals(arg))
-    if (!all(is_param)) {
-      txt <- if (sum(!is_param) == 1L) {
```
Comment (kaz462) [inline R/assertions.R:1080]: missed this for plural -
```R
> example_fun <- function(fun) {
+     assert_function(fun, params = c("x1", "X2"))
+ }
> 
> example_fun(assert_function)
Error in `example_fun()`:
! `x1` and `X2` is not an argument of the function specified for `fun`.
Run `rlang::last_trace()` to see where the error occurred.
```

**Review by bms63 — COMMENTED** [review]
Is there anyway to recommend to the user what they are missing?

image

For example, here I put in filter, but it should be filter_add?

Like a "Did you mean - filter_add" or if nothing else list out the possible arguments available to the function?

File: R/assertions.R [hunk 2]
```diff
@@ -1053,45 +1053,58 @@ assert_has_variables <- function(dataset, required_vars) {
...
+        ifelse(
+          !all(is_param),
+          "{enumerate(params[!is_param])} is not an argument
+           of the function specified for {.arg {arg_name}}.",
+          "{enumerate(params[!is_param])} are not arguments
+           of the function specified for {.arg {arg_name}}."
+        ),
+      call = call,
+      class = c(class, "assert-admiraldev")
+    )
```
Comment (kaz462) [inline R/assertions.R:1105]:
```suggestion
    if (!all(is_param)) {
      cli_abort(
        message = message %||%
          "{.val {params[!is_param]}} {?is/are} not {?an argument/arguments} 
        of the function specified for {.arg {arg_name}}.",
        call = call,
        class = c(class, "assert-admiraldev")
      )
    }
```
should we consider remove `enumerate` as well in final clean up?

Comment (bms63) [reply R/assertions.R:1105]: @kaz462 can you add this as issue in the super issue?  If that is your final comment, then can we merge in?

**Review by kaz462 — APPROVED** [review]
